module/zcommon/zfeature_common.c

```diff
@@ -559,6 +562,15 @@ zpool_feature_init(void)
 	    "com.datto:resilver_defer", "resilver_defer",
 	    "Support for deferring new resilvers when one is already running.",
 	    ZFEATURE_FLAG_READONLY_COMPAT, ZFEATURE_TYPE_BOOLEAN, NULL);
+
+	static const spa_feature_t zstd_deps[] = {
+		SPA_FEATURE_EXTENSIBLE_DATASET,
+		SPA_FEATURE_NONE
+	};
+	zfeature_register(SPA_FEATURE_ZSTD_COMPRESS,
+	    "org.freebsd:zstd_compress", "zstd_compress",
+	    "zstd compression algorithm support.",
+	    ZFEATURE_FLAG_PER_DATASET, ZFEATURE_TYPE_BOOLEAN, zstd_deps);
```

Comment:
Enabling `org.freebsd:zstd_compress` allow users to set `compression=zstd` on any dataset: this **does not activate** the feature but changes the MOS in a way that is incompatible with older versions (i.e. can be imported but will panic the system, even when imported "readonly"):

shell:
```
root@linux:~# cat /proc/sys/kernel/spl/gitrev
zfs-0.8.0-493-g2abfab9ed-dirty
root@linux:~# zpool import
   pool: testpool
     id: 14399741661690733518
  state: ONLINE
 action: The pool can be imported using its name or numeric identifier.
 config:

	testpool    ONLINE
	  sda       ONLINE
root@linux:~# zpool import -N -o readonly=on 14399741661690733518
[  392.289472] VERIFY(child < ZIO_COMPRESS_FUNCTIONS) failed
[  392.293082] PANIC at zio_compress.c:74:zio_compress_select()
```

kernel debugger:
```
Thread 132 hit Breakpoint 2, panic (fmt=0xffffffffc00ff330 "%s") at kernel/panic.c:73
73	kernel/panic.c: No such file or directory.
(gdb) bt
#0  panic (fmt=0xffffffffc00ff330 "%s") at kernel/panic.c:73
#1  0xffffffffc00f0493 in spl_panic (file=0xffffffffc04a2a50 "zio_compress.c", func=0xffffffffc0444e20 "zio_compress_select", line=74, fmt=0xffffffffc04afd5a "%s")
    at /usr/src/zfs/module/spl/../os/linux/spl/spl-err.c:70
#2  0xffffffffc0380a27 in zio_compress_select (spa=0xffff880035c18000, child=ZIO_COMPRESS_FUNCTIONS, parent=ZIO_COMPRESS_ON) at /usr/src/zfs/module/zfs/zio_compress.c:74
#3  0xffffffffc025f148 in compression_changed_cb (arg=0xffffffffc00ff330, newval=18446612137580132144) at /usr/src/zfs/module/zfs/dmu_objset.c:196
#4  0xffffffffc02b1758 in dsl_prop_register (ds=0xffff880035daa000, propname=<optimized out>, callback=<optimized out>, cbarg=<optimized out>) at /usr/src/zfs/module/zfs/dsl_prop.c:308
#5  0xffffffffc025f81e in dmu_objset_open_impl (spa=<optimized out>, ds=0xffff880035daa000, bp=<optimized out>, osp=0xffff880139df79d0) at /usr/src/zfs/module/zfs/dmu_objset.c:535
#6  0xffffffffc0263f43 in dmu_objset_from_ds (ds=0xffff880035daa000, osp=0xffff880139df7a80) at /usr/src/zfs/module/zfs/dmu_objset.c:663
#7  0xffffffffc0264121 in dmu_objset_hold_flags (name=<optimized out>, decrypt=B_FALSE, tag=0xffffffffc04359a4, osp=0xffff880139df7a80) at /usr/src/zfs/module/zfs/dmu_objset.c:701
#8  0xffffffffc0264176 in dmu_objset_hold (name=<optimized out>, tag=<optimized out>, osp=<optimized out>) at /usr/src/zfs/module/zfs/dmu_objset.c:713
#9  0xffffffffc02b01c4 in dsl_prop_get (dsname=<optimized out>, propname=0xffffffffc04afe43 "snapdev", intsz=8, numints=1, buf=0xffff880139df7ad8, setpoint=0x0 <irq_stack_union>)
    at /usr/src/zfs/module/zfs/dsl_prop.c:319
#10 0xffffffffc02b023e in dsl_prop_get_integer (ddname=<optimized out>, propname=<optimized out>, valuep=<optimized out>, setpoint=<optimized out>) at /usr/src/zfs/module/zfs/dsl_prop.c:342
#11 0xffffffffc0384987 in zvol_create_minors_cb (dsname=0xffffffffc00ff330 "%s", arg=0xffff880139df7730) at /usr/src/zfs/module/zfs/zvol.c:1041
#12 0xffffffffc02617a5 in dmu_objset_find_impl (spa=<optimized out>, name=0xffff880139df7b60 "", func=<optimized out>, arg=<optimized out>, flags=0) at /usr/src/zfs/module/zfs/dmu_objset.c:2865
#13 0xffffffffc02618ce in dmu_objset_find_impl (spa=<optimized out>, name=<optimized out>, func=<optimized out>, arg=<optimized out>, flags=0) at /usr/src/zfs/module/zfs/dmu_objset.c:2808
#14 0xffffffffc0261ba8 in dmu_objset_find (name=0xffff880139c3e000 "testpool", func=0xffffffffc0384930 <zvol_create_minors_cb>, arg=0xffff880139df7d10, flags=2) at /usr/src/zfs/module/zfs/dmu_objset.c:2881
#15 0xffffffffc038789d in zvol_create_minors_impl (name=<optimized out>) at /usr/src/zfs/module/zfs/zvol.c:1139
#16 zvol_task_cb (param=0xffff8800b3e8dc00) at /usr/src/zfs/module/zfs/zvol.c:1435
#17 0xffffffffc00f8f84 in taskq_thread (args=0xffff8800b7c6a7c0) at /usr/src/zfs/module/spl/../os/linux/spl/spl-taskq.c:936
#18 0xffffffff8109d03a in kthread (_create=0xffff8800ba979400) at kernel/kthread.c:211
#19 0xffffffff81809a8f in ret_from_fork () at arch/x86/entry/entry_64.S:468
#20 0x0000000000000000 in ?? ()
(gdb) 
```

I suggest adding `| ZFEATURE_FLAG_ACTIVATE_ON_ENABLE` to be on the safe side, or update the code to activate the feature when zstd is enabled on a dataset.

Reply:
The zstd_compress feature should be marked 'active' as soon as any dataset is using compress=zstd*

If that is not happening, that is definately a bug we need to address.

Reply:
@c0d3z3r0 I think at least adding `| ZFEATURE_FLAG_ACTIVATE_ON_ENABLE` is worthwhile, or the rewrite route, but I don't know how much rewrite effort that would require...

Reply:
I can confirm setting `compression=zstd` on a dataset does not activate the pool feature on my Debian builder running 3e723504b:
```
root@pull9735:~# cat /proc/sys/kernel/spl/gitrev
zfs-0.8.0-rc3-727-g3e723504b
root@pull9735:~# zpool create testpool /dev/sda -f
root@pull9735:~# zfs create testpool/fs -o compression=zstd
root@pull9735:~# zpool get all testpool | grep zstd
testpool  feature@zstd_compress          enabled                        local
root@pull9735:~# zfs get  compression testpool/fs
NAME         PROPERTY     VALUE           SOURCE
testpool/fs  compression  zstd            local
root@pull9735:~# 
```

Reply:
So we have, basically, two questions/issues with these flags:
1. If we enable ZSTD on a dataset, how do we want to act when importing said pool on a non-supporting version.
2. We should activate the pool feature on setting compression=zstd.

Reply:
We definately do NOT want ZFEATURE_FLAG_ACTIVATE_ON_ENABLE

This will make pools incompatible even when they do NOT contain any zstd compressed datasets.

On my test system, the feature flag went from 'enabled' to 'active' when I set compress=zstd on a dataset, and reverted to 'enabled' once that dataset was destroyed.

Reply:
I have just tested c48a09527 and the same thing is happening:
```
root@pull9024:~# cat /proc/sys/kernel/spl/gitrev
zfs-0.7.0-2625-gc48a09527
root@pull9024:~# zpool create testpool /dev/sda -f
root@pull9024:~# zfs create testpool/fs -o compression=zstd
root@pull9024:~# zpool get all testpool | grep zstd
testpool  feature@zstd_compress          enabled                        local
root@pull9024:~# zfs get  compression testpool/fs
NAME         PROPERTY     VALUE           SOURCE
testpool/fs  compression  zstd            local
root@pull9024:~#
```

I agree that changing the pool feature from enabled to active (and back) is the best option from a compatibility standpoint, but we should make sure this is actually happening. It's a bit worrisome we lost this functionality and only discovered it through manual testing, we should also update the ZTS to prevent this regression in the future.

Reply:
It seems you have to actually write a block for it to do what you expect.

```
# zpool create flagtest md1
# zpool get all flagtest|grep zstd
flagtest  feature@zstd_compress          enabled                        local
# zfs create -o compress=zstd flagtest/zstd
# zpool sync flagtest
# zpool get all flagtest | grep zstd
flagtest  feature@zstd_compress          enabled                        local
# cp /boot/kernel/kernel /flagtest/zstd/
# zpool sync flagtest
# zpool get all flagtest | grep zstd
flagtest  feature@zstd_compress          active                         local
# zfs destroy flagtest/zstd
# zpool get all flagtest | grep zstd
flagtest  feature@zstd_compress          enabled                        local
```

It seems the error you ran into was exporting and importing the pool without ever having written a block that uses zstd to activate the feature?

Reply:
Also, c48a095 is not part of this PR, you should probably test what is in this PR rather than random older versions

Reply:
> It seems the error you ran into was exporting and importing the pool without ever having written a block that uses zstd to activate the feature?

Yep, i run exactly the commands i reported in my comment:

```
cat /proc/sys/kernel/spl/gitrev
zpool create testpool /dev/sda -f
zfs create testpool/fs -o compression=zstd
zpool get all testpool | grep zstd
zfs get  compression testpool/fs
```

No data blocks compressed with zstd were ever written and the pool feature remains "enabled" and not "active": this creates a pool that can be imported (and will panic) systems running older bits. That is why i initially recommended `ZFEATURE_FLAG_ACTIVATE_ON_ENABLE` "to be on the safe side".

> Also, c48a095 is not part of this PR, you should probably test what is in this PR rather than random older versions

Though seemingly random, c48a095 is the latest commit pushed in #9024, which is the code i thought you were running on your test system. I also tested the latest commit in this PR on a dedicated builder which gave me the same results.

Reply:
@loli10K Please only discuss this PR here from now on. This PR is a refactor of Allan's code, we can NOT support Allans code here also.
Also please close this comment because it does not originate from this PR.

Reply:
> If you can, please repeat testing writing data to zstd on the lastest commit to this PR and check if the flag is set...

As i already said (https://github.com/zfsonlinux/zfs/pull/9735#discussion_r362106660, https://github.com/zfsonlinux/zfs/pull/9735#discussion_r362156472) i tested the latest commits from both PR, same bug.

Reply:
@loli10K 
Okey, but can you at least keep to best practices and make it possible for us to reproduce it?
Because Allan tried and didn't manage to trigger your bug.

Reply:
SHOOT, now I get it...
are you suggesting the MOS gets changed before data gets writhen to the zstd datasset?
In that case: Yes it would fail import before the flag is set.

Reply:
So, the ZFS 'feature flag' system, uses reference counting, and in general, it makes sense that we don't activate a feature until it is actually used.

However, in this case, we are tripping over the assert that the compression property has an out-of-bounds value on older versions of ZFS that don't know what ZSTD is.

It might not be possible to increase the refcount of the feature flag just because someone sets compress=zstd, because there isn't a good way to decrement the refcount in that case. However, I agree that having it end up causing a panic on a system is not good either.

We will need to experiment with whichever solution we come up with, and ensure that it is possible to return the feature flag to 'enabled' by destroying the final dataset that used zstd.

Reply:
> 
> 
> Though seemingly random, [c48a095](https://github.com/zfsonlinux/zfs/commit/c48a095278f404b2f98db9e8046b000ec058fcb1) is the latest commit pushed in #9024, which is the code i thought you were running on your test system. I also tested the latest commit in this PR on a dedicated builder which gave me the same results.

Sorry for the confusion. I have a branch here: https://github.com/allanjude/zfs/tree/freebsd_pr9735

which is this PR combined with the ZFS on FreeBSD branch, to run this PR basically unmodified on FreeBSD.

I also have a 2nd branch (freebsd_pr9735_kmem) of it that replaces the zstd_mempool allocator with the standard kmem_cache_alloc that works fine on FreeBSD, that I am testing to see if there is a performance difference.

Reply:
Looking into this more, the feature flag is activated in dsl_dataset_block_born(), the same as when activating a new hashing algorithm. So hashing algos likely have this same bug, where if you were to set checksum=skein, and then import the pool on an older version that did not support skein, you'd get the same ASSERT()

I am testing a change now to activate the dataset feature flag as soon as the user sets compress= or checksum=

Reply:
It's interesting to see how many issues this PR and its predecessors bring to light :P

Reply:
@allanjude Did you have any success or do you prefer @c0d3z3r0 to take a look?

Reply:
I have not yet managed to make the feature flag be 'activated' by setting compress=zstd, rather than by the first block actually being compressed with zstd.
I tried using compression_changed_cb() to trigger `dmu_objset_ds(os)->ds_feature_activation[f] = (void *)B_TRUE;` but it didn't seem to work.

I won't have time to work on this more in the next week, so please if anyone else has time to look into it.